utide/_time_conversion.py

```diff
@@ -2,44 +2,79 @@
 Utility for allowing flexible time input.
 """
 
-import warnings
+import numpy as np
 
-from datetime import date, datetime
 
+# to be added to get 1 on 1st January of year 1 from unix epoch '1970-01-01'
+_DAY_TO_GREGORIAN_EPOCH = 719163
 
-try:
-    from datetime import timezone
+# milisecond in a day
+_MS_PER_DAY = 1000 * 86400
 
-    have_tz = True
-except ImportError:
-    have_tz = False
+# datenum types
+_DATENUM_TYPES = (np.float, np.int64, np.int32)
 
 
-def _normalize_time(t, epoch):
-    if epoch == "python":
-        return t
-    if epoch == "matlab":
-        return t - 366
+def _date2num(date, epoch="1970-01-01 00:00:00.000"):
+    """
+    Numpy based date to datenum calculator.
+
+    `date` can be anything parsable by np.datetime64 - string, datetime, datetime64,
+    pandas datetime.
+
+    `epoch` is taken as the unix epoch.
+    """
+    date = np.atleast_1d(date)
+
     try:
-        epoch = datetime.strptime(epoch, "%Y-%m-%d")
-    except (TypeError, ValueError):
-        pass
-    if isinstance(epoch, date):
-        if not hasattr(epoch, "time"):
-            return t + epoch.toordinal()
-        # It must be a datetime, which is also an instance of date.
-        if epoch.tzinfo is not None:
-            if have_tz:
-                epoch = epoch.astimezone(timezone.utc)
+        date = date.astype("datetime64[ms]")
+    except ValueError:
+        raise ValueError("Cannot parse epoch as string or date or datetime")
+
+    try:
+        epoch = np.datetime64(epoch, "ms")
+    except ValueError:
+        raise ValueError("Cannot parse epoch as string or date or datetime")
+
+    # datenum calculation
+    datenum = (date - epoch).astype(float) / _MS_PER_DAY
+    return datenum
+
+
+def _python_gregorian_datenum(date):
+    """
+    Number of days since 0000-12-31.
+
+    Python gregorian time is 1 on 1st day of 1st year. Essentially, it means,
+    the epoch for python gregorian time is 0000-12-31. With _date2num() defined
+    above, this amounts to 719163 days from the unix-epoch 1970-01-01 00:00:00.
+    To avoid repetative calculation, this is defined as _DAY_TO_GREGORIAN_EPOCH.
+    """
+    return _date2num(date) + _DAY_TO_GREGORIAN_EPOCH
+
+
+def _normalize_time(t, epoch=None):
+    """
+    Convert datetime or datenum array to proper input datenum array with an
+    epoch from '0000-12-31' - 1st Jan of 1st year is 1.
```

Comment:
I think we could just as well normalize to the unix epoch; no need to go back and forth all the way to '000-12-31'.

Reply:
Yes, that would probably be ideal. In that case, the astronomic calculation in ut_astron needs to be changed, isn't it (the `daten` variable)? Any other known place where it could impact? As you might have already guessed, in this proposed change, my goal was to touch the codebase as minimum as possible.

Reply:
My recollection from looking at this some time ago is that ut_astron is the only function that depends on the "python" epoch.  It's fully public, even though it seems unlikely that anyone is depending on it externally.  Options:

1. Deprecate it in favor of a nearly identical function with a different name (just plain "astron"? "astron_vars"?) that takes time in the unix epoch.  `ut_astron` could be stripped down to a wrapper for the new function.
2. Add an epoch kwarg to the existing version, defaulting to the python epoch, but set to the unix epoch when we call the function internally.

The second option seems like the easiest, and quite satisfactory.  Your thoughts?

As far as the implementation goes I would not change the function's *internal* use of yet another epoch, "1899-12-31 at noon".

Reply:
Also, this change could be done in a separate PR.